src/peft/utils/other.py

```diff
@@ -481,3 +481,80 @@ def id_tensor_storage(tensor: torch.Tensor) -> Tuple[torch.device, int, int]:
         unique_id = storage_ptr(tensor)
 
     return tensor.device, unique_id, storage_size(tensor)
+
+
+def separate_pad_tokens(
+    input_ids: torch.Tensor, inputs_embeds: torch.Tensor, attention_mask: torch.Tensor, labels: torch.Tensor
+):
+    """
+    Separates out the tokens corresponding to padding in input_ids, inputs_embeds, attention_mask and labels Returns:
+    - input_ids: input_ids with padding tokens removed
+    - inputs_embeds: inputs_embeds with padding tokens removed
+    - attention_mask: attention_mask with padding tokens removed
+    - labels: labels with padding tokens removed
+    - pad_els: a tuple (pad_token_input_ids, pad_token_inputs_embeds, pad_token_attention_mask, pad_token_labels)
+      containing the padding token elements
+    """
+    if input_ids is None and inputs_embeds is None:
+        raise ValueError("You have to provide either input_ids or inputs_embeds")
+    ndim = input_ids.ndim if input_ids is not None else inputs_embeds.ndim - 1
+    if not ndim == 2:
+        raise ValueError("input_ids must be a 2D tensor")
+    # find the first index of a non-padding token
+    if attention_mask is not None:
+        first_non_padding_tok_ind = torch.argmax(attention_mask, dim=1)
+    else:
+        batch_size = _get_batch_size(input_ids, inputs_embeds)
+        first_non_padding_tok_ind = torch.zeros(batch_size, dtype=torch.long)
```

Comment:
Could you please explain why we can make this assumption? Also, does this mean that if `attention_mask` is `None`, all the special handling could be skipped?

Reply:
This logic is again because this fix is at the model stage. If `attention_mask` is `None`, then, without knowing the padding token id, there's no way to segregate different input tokens in `input_ids`. So the assumption is that there is no padding done. One more alternative to this change is to change the API itself for prompt tuning: ask the user to always pass `pad_token_id`, and raise an error if not. This is a breaking change though so I didn't go for this.

The special handling can indeed be completely skipped if `attention_mask` is `None`. My issue is that it makes the types more unclear that it is already. Currently, if it's prompt tuning, then all the input tensors are internally passed around as `List[torch.Tensor]` until the final `model.forward` call. If I skip the special handling completely, then it will sometimes be `List[torch.Tensor]` and sometimes `torch.Tensor` internally. Let me see what I can do.

Reply:
> without knowing the padding token id, there's no way to segregate different input tokens in `input_ids`.

Okay, so this is basically saying: If there is no attention mask, we can't do anything, so we throw our hands in the air :) And not necessarily that this is always the correct behavior.

> This is a breaking change though so I didn't go for this.

Yeah, that's a tricky problem, as we want to indeed avoid breaking existing code. If the existing code were doing a totally incorrect thing, with the result being nonsense, I could see an argument for raising an error. But as you mentioned earlier, it's not a given that the result would be nonsense.

> The special handling can indeed be completely skipped if `attention_mask` is `None`. My issue is that it makes the types more unclear that it is already.

I see the argument. Maybe if we could do something like returning early on the corresponding `PeftModelForCausalLM` methods (as we do for non-prompt-learning methods), we could avoid having a bunch of if-else for list of tensors vs tensors. WDYT?

Reply:
Yes that sounds good! I'll refactor that way.